### PR TITLE
[RFC] Rewrite the picker implementation

### PR DESCRIPTION
**Pull request by aschempp**
As discussed in several issues, the current picker implementation does not really work. Here's my rewrite to support all current features and some more!

**TODO:**
- [x] Improve method names
- [x] Fix coding style / docblocks
- [x] Add necessary interfaces
- [x] Add unit tests

### CODE REVIEW

**Comment by leofeyer** [comment]
Please revert the naming and namespace changes. Otherwise it is impossible to understand the changes.

File: src/Resources/contao/templates/backend/be_tinyNews.html5 [hunk 1]
```diff
@@ -24,6 +33,7 @@ setTimeout(function() {
       }
       editor.on('focus', function(){ Backend.getScrollOffset(); });
     },
+    file_browser_callback_types: <?= json_encode($pickerTypes) ?>,
```
Comment (leofeyer) [inline src/Resources/contao/templates/backend/be_tinyNews.html5]: The whole logic needs to be moved to the `DataContainer` class, so we can use `$this->pickerTypes` in the templates.

File: src/Resources/contao/classes/DataContainer.php [hunk 2]
```diff
@@ -510,7 +514,19 @@ protected function row($strPalette=null)
 			}
 		}
 
-		$objWidget->wizard = $wizard;
+		if ($wizard != '')
```
Comment (leofeyer) [inline src/Resources/contao/classes/DataContainer.php]: How is this change related to the picker?

File: src/Resources/public/core.js [hunk 3]
```diff
@@ -898,7 +898,7 @@ var Backend =
 		var opt = options || {},
 			maxWidth = (window.getSize().x - 20).toInt(),
 			maxHeight = (window.getSize().y - 192).toInt();
-		if (!opt.id) opt.id = 'tl_select';
+		if (!opt.id) opt.id = 'tl_listing';
```
Comment (leofeyer) [inline src/Resources/public/core.js]: You must not change this, because it is not backwards compatible with the old `PageTree` and `FileTree` widgets.

File: src/Resources/contao/classes/DataContainer.php (hunk 2 above)
Comment (aschempp) [inline src/Resources/contao/classes/DataContainer.php]: It hides the picker icon and spacing if the dcaPicker is not supported (due to permissions etc.)

File: src/Resources/contao/templates/backend/be_tinyNews.html5 (hunk 1 above)
Comment (aschempp) [inline src/Resources/contao/templates/backend/be_tinyNews.html5]: Is tinyMCE initialized by DataContainer ?

Comment (leofeyer) [inline src/Resources/contao/templates/backend/be_tinyNews.html5]: Yes: https://github.com/contao/core-bundle/blob/master/src/Resources/contao/classes/DataContainer.php#L537-L551

Comment (aschempp) [inline src/Resources/contao/templates/backend/be_tinyNews.html5]: Changed in 1f24de033a5f8fc4a3739ff8e6e8c53b1507f5ed

File: src/Picker/PickerBuilder.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            $this->menuFactory,
+            $providers,
+            $config
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createFromJson($json)
```
Comment (Toflar) [inline src/Picker/PickerBuilder.php]: I would actually add `createFromUrlValue()` and `saveForUrlValue()` or something like this. You can keep the others if you like but this would allow us using `gzencode()` etc. in these methods (or anything else in the future) to transparently and backwards compatible improve the size of the content :)

File: src/Resources/contao/drivers/DC_Table.php [hunk 5]
```diff
@@ -6135,17 +6135,13 @@ public function initPicker(PickerInterface $picker)
...
-				if (!empty($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root']))
+				if (!empty($this->root))
 				{
-					$arrRoot = array_intersect($arrRoot, $GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root']);
-
-					if (!empty($arrRoot))
-					{
-						$arrRoot = $this->eliminateNestedPages($arrRoot);
-					}
+					$arrRoot = array_intersect($arrRoot, $this->Database->getChildRecords($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root'], 'tl_page'));
```
Comment (Toflar) [inline src/Resources/contao/drivers/DC_Table.php]: This should be `$this->strTable` not `tl_page` right?

Comment (aschempp) [inline src/Resources/contao/drivers/DC_Table.php]: Fixed in 368b144dd3e88de3e803cb728a5cf13aa97332fe

File: src/Picker/PickerBuilder.php (hunk 4 above)
Comment (aschempp) [inline src/Picker/PickerBuilder.php]: implemented in a25b13229627dfc085db24115786da8499ec97dc

Comment (aschempp) [inline src/Picker/PickerBuilder.php]: implemented in a25b13229627dfc085db24115786da8499ec97dc, including Gzip compression

File: src/Controller/BackendController.php [hunk 6]
```diff
@@ -224,8 +226,23 @@ public function alertsAction()
      */
     public function pickerAction(Request $request)
```
Comment (Toflar) [inline src/Controller/BackendController.php]: Can we have some extended description in the docblock here? :) `Handles the picker redirect.` does not really help a lot.

File: src/Picker/AbstractPickerProvider.php [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+     */
+    private $tokenStorage;
+
+    /**
+     * Constructor.
+     *
+     * @param FactoryInterface      $menuFactory
+     * @param TokenStorageInterface $tokenStorage
+     */
+    public function __construct(FactoryInterface $menuFactory, TokenStorageInterface $tokenStorage)
```
Comment (Toflar) [inline src/Picker/AbstractPickerProvider.php]: I prefer having setter injection for the `TokenStorageInterface` because `getUser()` is absolutely optional for an `AbstractPickerProvider`. I think we should have `setTokenStorage()` and `getTokenStorage()`. The `getUser()` already considers the `null` case anyway :)

File: src/Picker/DcaPickerProviderInterface.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\Picker;
+
+/**
+ * Interface for DCA picker providers.
```
Comment (Toflar) [inline src/Picker/DcaPickerProviderInterface.php]: I would like to have more description here. `Use this if .....` or something. This would help very much.

File: src/Picker/PickerConfig.php [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /**
+     * Encodes the picker configuration for use in the URL.
+     *
+     * @param bool $gzip
+     *
+     * @return string
+     */
+    public function urlEncode($gzip = true)
```
Comment (Toflar) [inline src/Picker/PickerConfig.php]: I would name this `$compress = true`. Who cares if this is gzip.

File: src/Resources/contao/drivers/DC_Folder.php [hunk 10]
```diff
@@ -3042,67 +3037,58 @@ protected function isProtectedPath($path)
...
 	{
-		parent::setDcaFilter($arrFilter);
+		$attributes = parent::initPicker($picker);
 
-		$blnHideFiles = !isset($arrFilter['files']) && !isset($arrFilter['filesOnly']);
+		$this->blnFiles = false;
+		$this->blnFilesOnly = false;
 
-		if (isset($arrFilter['files']) && $arrFilter['files'] === false)
+		if (is_array($attributes))
```
Comment (Toflar) [inline src/Resources/contao/drivers/DC_Folder.php]: I'd prefer a `null` check as there can only be array or null.

File: src/Resources/contao/drivers/DC_Table.php [hunk 11]
```diff
@@ -6123,4 +6119,32 @@ protected function formatGroupHeader($field, $value, $mode, $row)
...
 	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function initPicker(PickerInterface $picker)
+	{
+		$attributes = parent::initPicker($picker);
+
+		if (is_array($attributes))
```
Comment (Toflar) [inline src/Resources/contao/drivers/DC_Table.php]: Check for `null` here and do an early return please. I hate these 5-level nestings 😄

File: src/Picker/AbstractPickerProvider.php (hunk 7 above)
Comment (leofeyer) [inline src/Picker/AbstractPickerProvider.php]: I think this makes sense.

File: src/Picker/ArticlePickerProvider.php [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+            $attributes['value'] = str_replace(['{{article_url::', '}}'], '', $config->getValue());
+        }
+
+        return $attributes;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function convertDcaValue(PickerConfig $config, $value)
```
Comment (leofeyer) [inline src/Picker/ArticlePickerProvider.php]: We are actually converting the picked value here, aren't we?

File: src/Picker/PickerConfig.php [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+            $this->extras,
+            $this->value,
+            $current
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function jsonSerialize()
```
Comment (leofeyer) [inline src/Picker/PickerConfig.php]: The method says JSON serialize but it returns an array?

File: src/Picker/PickerConfig.php [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    /**
+     * Encodes the picker configuration for use in the URL.
+     *
+     * @param bool $gzip
+     *
+     * @return string
+     */
+    public function urlEncode($gzip = true)
+    {
+        $data = json_encode($this);
```
Comment (leofeyer) [inline src/Picker/PickerConfig.php]: Why are you encoding `$this` here? Should it be `$this->jsonSerialize()`?

File: src/Picker/PickerInterface.php [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+     * @return ItemInterface
+     */
+    public function getMenu();
+
+    /**
+     * Gets the current provider.
+     *
+     * @return PickerProviderInterface|null
+     */
+    public function getCurrentProvider();
```
Comment (leofeyer) [inline src/Picker/PickerInterface.php]: I would prefer `getProvider()` here.

File: src/Picker/PickerInterface.php [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * Gets URL to the current picker tab.
+     *
+     * @return string
+     */
+    public function getCurrentUrl();
```
Comment (leofeyer) [inline src/Picker/PickerInterface.php]: I would prefer `getUrl()` here.

File: src/Picker/PickerProviderInterface.php [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    public function supportsValue(PickerConfig $config);
+
+    /**
+     * Returns whether the picker is currently active.
+     *
+     * @param PickerConfig $config
+     *
+     * @return bool
+     */
+    public function isCurrent(PickerConfig $config);
```
Comment (leofeyer) [inline src/Picker/PickerProviderInterface.php]: I would prefer `isActive()` here.

File: src/Resources/contao/classes/DataContainer.php [hunk 18]
```diff
@@ -558,17 +564,26 @@ protected function row($strPalette=null)
...
 			$objTemplate->type = $type;
+			$objTemplate->fileBrowserTypes = $fileBrowserTypes;
 
 			// Deprecated since Contao 4.0, to be removed in Contao 5.0
 			$objTemplate->language = \Backend::getTinyMceLanguage();
 
 			$updateMode = $objTemplate->parse();
 
-			unset($file, $type);
+			unset($file, $type, $pickerBuilder);
```
Comment (leofeyer) [inline src/Resources/contao/classes/DataContainer.php]: Also unset `$fileBrowserTypes`?

File: src/Resources/contao/controllers/BackendMain.php [hunk 19]
```diff
@@ -113,7 +113,22 @@ public function run()
 		// Open a module
 		elseif (\Input::get('do'))
 		{
-			$this->Template->main .= $this->getBackendModule(\Input::get('do'));
+			$picker = null;
+
+			if (isset($_GET['picker']))
```
Comment (leofeyer) [inline src/Resources/contao/controllers/BackendMain.php]: Using `isset($_GET)` cannot be "overwritten" via `Input::setGet()`. I know that I have been doing it, too – maybe we should discuss it and be consistent?

File: src/Resources/contao/drivers/DC_Folder.php [hunk 20]
```diff
@@ -2667,7 +2662,7 @@ protected function generateTree($path, $intMargin, $mount=false, $blnProtected=t
 			}
 		}
 
-		if ($this->blnHideFiles)
+		if (!$this->blnFiles && !$this->blnFilesOnly)
```
Comment (leofeyer) [inline src/Resources/contao/drivers/DC_Folder.php]: Does this really work without `$this->strPickerField &&`?

File: src/Resources/contao/templates/backend/be_main.html5 [hunk 21]
```diff
@@ -94,7 +94,7 @@
 
   <?php if ($this->isPopup): ?>
     <script>
-      if ($$('.tl_tree_checkbox,.tl_tree_radio').length > 0) {
+      if ($$('.tl_tree_checkbox,.tl_tree_radio').length > 0 && !document.location.search.test('act=select')) {
```
Comment (leofeyer) [inline src/Resources/contao/templates/backend/be_main.html5]: I'd prefer to have this as an outer if condition:

```js
if (!document.location.search.test('act=select')) {
    if ($$('.tl_tree_checkbox,.tl_tree_radio').length > 0)) {
        // …
    } else {
        // …
    }
}
```

Otherwise the else condition would still be evaluated in "edit multiple" mode.

**Comment by leofeyer** [comment]
Already looking very good! Thank you very much @aschempp.

File: src/Controller/BackendController.php (hunk 6 above)
Comment (aschempp) [inline src/Controller/BackendController.php]: Improved in 9e53148bafbf41b76aef3e407678d25129765cb4

File: src/Picker/DcaPickerProviderInterface.php (hunk 8 above)
Comment (aschempp) [inline src/Picker/DcaPickerProviderInterface.php]: Improved in 9e53148bafbf41b76aef3e407678d25129765cb4

File: src/Picker/PickerConfig.php (hunk 9 above)
Comment (aschempp) [inline src/Picker/PickerConfig.php]: Improved in 9e53148bafbf41b76aef3e407678d25129765cb4

File: src/Resources/contao/drivers/DC_Folder.php (hunk 10 above)
Comment (aschempp) [inline src/Resources/contao/drivers/DC_Folder.php]: Improved in 9e53148bafbf41b76aef3e407678d25129765cb4

File: src/Resources/contao/drivers/DC_Table.php (hunk 11 above)
Comment (aschempp) [inline src/Resources/contao/drivers/DC_Table.php]: Improved in 9e53148bafbf41b76aef3e407678d25129765cb4

File: src/Picker/ArticlePickerProvider.php (hunk 12 above)
Comment (aschempp) [inline src/Picker/ArticlePickerProvider.php]: We're converting the value from DC/database to the one we can pick.

File: src/Picker/PickerConfig.php (hunk 13 above)
Comment (aschempp) [inline src/Picker/PickerConfig.php]: That's how the `\JsonSerializable` interface works 😉

File: src/Picker/PickerConfig.php (hunk 14 above)
Comment (aschempp) [inline src/Picker/PickerConfig.php]: That's how the `\JsonSerializable` interface works 😉

File: src/Picker/PickerInterface.php (hunk 15 above)
Comment (aschempp) [inline src/Picker/PickerInterface.php]: It returns the currently active provider (based on the value).

File: src/Picker/PickerInterface.php (hunk 16 above)
Comment (aschempp) [inline src/Picker/PickerInterface.php]: It returns the URL for the current provider (based on the value).

File: src/Picker/PickerProviderInterface.php (hunk 17 above)
Comment (aschempp) [inline src/Picker/PickerProviderInterface.php]: I've named this after the `current` variable of the `KnpMenu`…

File: src/Resources/contao/classes/DataContainer.php (hunk 18 above)
Comment (aschempp) [inline src/Resources/contao/classes/DataContainer.php]: Added in 9e53148bafbf41b76aef3e407678d25129765cb4

File: src/Resources/contao/drivers/DC_Folder.php (hunk 20 above)
Comment (aschempp) [inline src/Resources/contao/drivers/DC_Folder.php]: Yes it does, because `$blnFiles` is [initialized to `true`](https://github.com/aschempp/contao-core-bundle/blob/6a7d052a4b08047e4d42cd9f084d27d195aea0b1/src/Resources/contao/drivers/DC_Folder.php#L84) and only [set to `false` on `initPicker`](https://github.com/aschempp/contao-core-bundle/blob/6a7d052a4b08047e4d42cd9f084d27d195aea0b1/src/Resources/contao/drivers/DC_Folder.php#L3046).

File: src/Resources/contao/templates/backend/be_main.html5 (hunk 21 above)
Comment (aschempp) [inline src/Resources/contao/templates/backend/be_main.html5]: That would mean the buttons are not disabled when in select mode, which is not what we want, right?

File: src/Picker/AbstractPickerProvider.php (hunk 7 above)
Comment (aschempp) [inline src/Picker/AbstractPickerProvider.php]: Changed in 37b9d7caa43b196d8b234d59440c1b09f3842e06

File: src/Picker/ArticlePickerProvider.php (hunk 12 above)
Comment (leofeyer) [inline src/Picker/ArticlePickerProvider.php]: I still prefer `convertValue()` or `convertPickedValue()`. 😄

File: src/Picker/PickerInterface.php (hunk 15 above)
Comment (leofeyer) [inline src/Picker/PickerInterface.php]: Then I prefer `getActiveProvide()`.

File: src/Picker/PickerConfig.php (hunk 14 above)
Comment (leofeyer) [inline src/Picker/PickerConfig.php]: This is actually because `json_encode()` can handle objects and not because of the interface. 😉

File: src/Picker/PickerInterface.php (hunk 15 above)
Comment (aschempp) [inline src/Picker/PickerInterface.php]: It's named after `isCurrent`, which is based on `KnpMenu` (as described below). Still wanna change it?

File: src/Picker/PickerProviderInterface.php (hunk 17 above)
Comment (leofeyer) [inline src/Picker/PickerProviderInterface.php]: Hm, ok. We should either say `current` or `active` everywhere for consistency. So `getCurrentProvider()`, `getCurrentUrl()`, `isCurrent()` etc. or `getActiveProvider()`, `getActiveUrl()`, `isActive()` etc.

I prefer the latter, but as long as it is consistent, I can live with it.

File: src/Picker/ArticlePickerProvider.php (hunk 12 above)
Comment (aschempp) [inline src/Picker/ArticlePickerProvider.php]: You know the method is (only) part of the `DcaPickerProviderInterface`, right? That's why I named the methods so they are specific to the DCA, if we ever later add similar features for something else…

File: src/Picker/PickerConfig.php (hunk 14 above)
Comment (aschempp) [inline src/Picker/PickerConfig.php]: Well `json_encode` will realize the object implements `\JsonSerializable` and uses that method…

File: src/Resources/contao/controllers/BackendMain.php (hunk 19 above)
Comment (leofeyer) [inline src/Resources/contao/controllers/BackendMain.php]: You are right. :thumbsup:

File: src/Resources/contao/drivers/DC_Folder.php (hunk 20 above)
Comment (leofeyer) [inline src/Resources/contao/drivers/DC_Folder.php]: Ok.

File: src/Picker/PickerProviderInterface.php (hunk 17 above)
Comment (aschempp) [inline src/Picker/PickerProviderInterface.php]: Yeah, sort of agree. We name it *active* most of the time in Contao. But we cannot change `KnpMenu`, and the `isCurrent()` method is also called in https://github.com/aschempp/contao-core-bundle/blob/37b9d7caa43b196d8b234d59440c1b09f3842e06/src/Picker/AbstractPickerProvider.php#L64 to determine the current menu item…

File: src/Resources/contao/templates/backend/be_main.html5 (hunk 21 above)
Comment (leofeyer) [inline src/Resources/contao/templates/backend/be_main.html5]: Of course. Don't know what I was thinking. 😄

File: src/Picker/ArticlePickerProvider.php (hunk 12 above)
Comment (leofeyer) [inline src/Picker/ArticlePickerProvider.php]: I just think that `convertPickedValue()` is better understandable for everyone implementing the interface. `convertDcaValue()` sounds like the `load_callback` of the field is executed to convert the DCA value. 😉

Comment (aschempp) [inline src/Picker/ArticlePickerProvider.php]: How about `convertDcaToPickerValue()`?

File: src/Picker/PickerProviderInterface.php (hunk 17 above)
Comment (leofeyer) [inline src/Picker/PickerProviderInterface.php]: Just use "current" everywhere.

Comment (aschempp) [inline src/Picker/PickerProviderInterface.php]: 👍  so nothing to change 😎

File: src/Picker/ArticlePickerProvider.php (hunk 12 above)
Comment (leofeyer) [inline src/Picker/ArticlePickerProvider.php]: What bothers you about `convertPickedValue()`?

Comment (aschempp) [inline src/Picker/ArticlePickerProvider.php]: Every other method in the `DcaPickerProviderInterface` contains `Dca` 🙈

Comment (leofeyer) [inline src/Picker/ArticlePickerProvider.php]: Ok, first, all the `Dca` parts are unnecessary in my opinion, because the class already has `Dca` in it.

But second, the first two methods actually read data from the DCA, so using `Dca` in the method name is kind of correct. The third method, however, reads user input, so …

Comment (aschempp) [inline src/Picker/ArticlePickerProvider.php]: You mean the interface does?
The `convertDcaValue` does not really read user input, it's basically a save_callback on the DCA value before it is rendered as a picker radio/checkbox value.

Comment (leofeyer) [inline src/Picker/ArticlePickerProvider.php]: Of course it reads user input. The user picks page ID 12, submits the value and gets back `{{link_url::12}}`.

Again, I think that `convertPickedValue()` is much better understandable.

Comment (aschempp) [inline src/Picker/ArticlePickerProvider.php]: That is not correct. There is no Ajax request anymore, the values are converted **before** the user selects them in the picker. This also results in an instant update of the field value, contrary to the previous implementation which took seconds to update 😉

Comment (leofeyer) [inline src/Picker/ArticlePickerProvider.php]: Ah, why don't you say this right away? Then `convertDcaValue` is actually fine. 😄

**Comment by leofeyer** [comment]
Actually merged in 874076de55cbf962fd5e41c00e853e4dec8842a6.
